edb/pgsql/compiler/relctx.py

```diff
@@ -749,9 +749,13 @@ def create_iterator_identity_for_path(
         if apply_volatility:
             apply_volatility_ref(stmt, ctx=ctx)
 
-    pathctx.put_path_var(stmt, path_id, id_expr, force=True, aspect='iterator')
     pathctx.put_path_bond(stmt, path_id, iterator=True)
 
+    if (path_id, 'iterator') in stmt.path_rvar_map:
+        return
+
+    pathctx.put_path_var(stmt, path_id, id_expr, aspect='iterator')
```

Comment:
@msullivan I'm not sure if this is the correct way to do it.

I had a problem where DML within iteration was producing this:
```SQL
    WITH
    source_rel AS
        ((
            SELECT *, ctid FROM "the_link_table"
            -- this ctid should be used as the iterator
        )),
    "iter~1" AS
        ((
            FROM source_rel AS "source_rel~1"
            CROSS JOIN LATERAL (
                SELECT/*<pg.SelectStmt at 0x7f82c0915f10>*/
                    "source_rel~1"."hell" AS "hell_value~1",
                    edgedb.uuid_generate_v4() AS "hell_iterator~1"
                    -- this uuid_generate call should not be here.
                    -- ctid should be used instead.
                    -- I inject its path and rel vars so it would be picked up
                    -- but the code above overriddes it.
                    -- So I've disabled that.
                WHERE
                    ("source_rel~1"."hell" IS NOT NULL)
            ) AS "Hello~3"
            SELECT/*<pg.SelectStmt at 0x7f82c0914110>*/
                "Hello~3"."hell_iterator~1" AS "hell_iterator~2"
        )),
    "ins_contents~1" AS
        (...)
    ...
```

Reply:
This seems plausible, though I'll ponder it a bit.
(I would restructure it so that the put_path_var is in the conditional instead of having an early return.)